doc/compiler/il/GlRegDeps.md

```diff
@@ -0,0 +1,152 @@
+<!--
+Copyright (c) 2021, 2021 IBM Corp. and others
+
+This program and the accompanying materials are made available under
+the terms of the Eclipse Public License 2.0 which accompanies this
+distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+or the Apache License, Version 2.0 which accompanies this distribution and
+is available at https://www.apache.org/licenses/LICENSE-2.0.
+
+This Source Code may also be made available under the following
+Secondary Licenses when the conditions for such availability set
+forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+General Public License, version 2 with the GNU Classpath
+Exception [1] and GNU General Public License, version 2 with the
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# Global Register Dependency
+
+`GlRegDeps` at its core is really a code gen register dependency. It associates virtual registers with real registers, or pairs virtual registers with real registers.
+
+Register dependencies enforce constraints. It says that this value has to be in this register at this time. Global register dependency is fundamentally about conveying particular values in particular registers across the block boundaries. The values and registers are expressed by `GlRegDeps` from one basic block to another.
+
+Most common control flow opcodes are: `if`, `goto`, `return`, `athrow`, `switch`. They should be the last tree before the `BBEnd`. They should have their own `GlRegDeps` for the branch target if required. The `GlRegDeps` under `BBEnd` is for the fall through case. It is important to make sure it has the right value in the right real register at the point where the `GlRegDeps` is evaluated.
+
+If a value that is expected to be in a register is not in the register, or the register is not preserved if it is a call, it leads to register shuffling. The local register allocator has to shuffle it to a different register to make sure it has the right value in `GlRegDeps`. Register shuffling does come at a cost. It should be avoided when possible.
+
+GRA related opcodes such as `GlRegDeps` use the reference count to determine when a register is no longer needed. It is a basic rule within the codegen for nodes and their associated virtual registers in general. It has to be set up correctly, otherwise it could result in registers being overwritten with some other values prematurely. In the following example, `n190n` is referenced `5` times but its reference count is set up incorrectly as `6` (`rc=6`).
```

Comment:
This is the first time we use the term "GRA" so we should spell it out so as to avoid reader confusion on the acronym, i.e.:

> Global Register Allocation (GRA)

Reply:
Updated in 9c4e378